python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,486 @@
+        
+        TestPubSub.publish_and_assert(all_clients)
+
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    async def test_pubsub_pattern_and_exact(self, request, cluster_mode):
+        """ Tests that subscription to exact channel and it's pattern triggers reception of 2 messages """
+
+        CLIENTS_COUNT = 5
+        CHANNEL_NAME = "test-channel"
+        CHANNEL_PATTERN = "test-*"
+        MESSAGES_BY_CLIENT = {
+            id : {
+                CHANNEL_NAME: {
+                    msg + str(id): 1 for msg in ["foo_from_", "bar_from_", "baz_from_"]
+                }
+            } for id in range(CLIENTS_COUNT)
+        }
+
+        all_clients: List[TestPubSub.PubSubClient] = []
+        for id in range(CLIENTS_COUNT):
+            all_clients.append(TestPubSub.PubSubClient(
+                subsciptions={
+                    CoreCommands.ChannelModes.Pattern: {
+                        CHANNEL_PATTERN
+                    },
+                    CoreCommands.ChannelModes.Exact: {
+                        CHANNEL_NAME
+                    }
+                },
+                sharded_messages_to_publish=[],
+                unsharded_messages_to_publish=MESSAGES_BY_CLIENT[id],
+                cluster_mode=cluster_mode,
+                request=request))
+            all_clients[-1].connect_and_subscribe()
+        
+        TestPubSub.publish_and_assert(all_clients)
+
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    async def test_pubsub_compound_pattern(self, request, cluster_mode):
+        """ Tests that subscription to compound patterns triggers reception of 2 messages (one for each pattern)  """
+
+        CLIENTS_COUNT = 5
+        CHANNEL_NAME = "test-channel"
+        CHANNEL_PATTERNS = {"test-*", "test?channel"}
+        MESSAGES_BY_CLIENT = {
+            id : {
+                CHANNEL_NAME: {
+                    msg + str(id): 1 for msg in ["foo_from_", "bar_from_", "baz_from_"]
+                }
+            } for id in range(CLIENTS_COUNT)
+        }
+
+        all_clients: List[TestPubSub.PubSubClient] = []
+        for id in range(CLIENTS_COUNT):
+            all_clients.append(TestPubSub.PubSubClient(
+                subsciptions={
+                    CoreCommands.ChannelModes.Pattern: CHANNEL_PATTERNS
+                },
+                sharded_messages_to_publish=[],
+                unsharded_messages_to_publish=MESSAGES_BY_CLIENT[id],
+                cluster_mode=cluster_mode,
+                request=request))
+            all_clients[-1].connect_and_subscribe()
+        
+        TestPubSub.publish_and_assert(all_clients)
+
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("sharded_pubsub", [True, False])
+    async def test_pubsub_channel_boundary_exact(self, request, cluster_mode, sharded_pubsub):
+        """ Tests that messages do not cross channel boundaries by exact subsctiption """
```

Comment:
I don't understand what this test does

Reply:
clients listening on "channel-x" wont get messages published on "channel=y"

Reply:
what bugs do you expect this test to catch?

Reply:
Bugs that may lead to leakage of messages across the channels. This is something we have to test.